src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,158 @@
+    private Node<T> last;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value);
+        if (isEmpty()) {
+            first = newNode;
+            last = newNode;
+        } else {
+            newNode.prev = last;
+            last.next = newNode;
+            last = newNode;
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size()) {
+            throw new ArrayIndexOutOfBoundsException("Index: " + index + ", Size: " + size());
+        }
+
+        if (index == size()) {
+            add(value);
+            return;
+        }
+
+        Node<T> newNode = new Node<>(value);
+        Node<T> current = getNode(index);
+
+        if (current.prev == null) {
+            first = newNode;
+        } else {
+            Node<T> previous = current.prev;
+            previous.next = newNode;
+            newNode.prev = current.prev;
+        }
+
+        newNode.next = current;
+        current.prev = newNode;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNode(index).data;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> node = getNode(index);
+        T oldValue = node.data;
+        node.data = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> nodeToRemove = getNode(index);
+        unlink(nodeToRemove);
+        return nodeToRemove.data;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = first;
+        while (current != null) {
+            if (current.data == null ? object == null : current.data.equals(object)) {
+                unlink(current);
+                return true;
+            }
+
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        int count = 0;
+        Node<T> current = first;
+        while (current != null) {
+            count++;
+            current = current.next;
+        }
+        return count;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return first == null;
+    }
+
+    private static class Node<T> {
+        private T data;
+        private Node<T> prev;
+        private Node<T> next;
+
+        Node(T data) {
+            this.data = data;
+        }
+
+        // Геттеры и сеттеры для prev и next
+        public Node<T> getPrev() {
+            return prev;
+        }
+
+        public void setPrev(Node<T> prev) {
+            this.prev = prev;
+        }
+
+        public Node<T> getNext() {
+            return next;
+        }
+
+        public void setNext(Node<T> next) {
+            this.next = next;
+        }
+    }
```

Comment:
![image](https://github.com/mate-academy/jv-linked-list/assets/63095184/012aa16f-e349-425f-8276-afb6f80646ab)
Better place for Node in the end of class